### PR TITLE
Finalized resume components

### PR DESCRIPTION
**Pull request by AshleyLuEllen**

### CODE REVIEW

File: src/main/java/to/us/resume_builder/resume_components/category/BulletCategory.java [hunk 1]
```diff
@@ -1,35 +1,83 @@
...
         return id;
     }
 
-    // TODO Implement
+    /**
+     * Removes the list item that matches the id
+     * @param id
+     */
     public void removeBullet(String id){
+        bullets = bullets.stream().filter(e -> !e.getId().equals(id)).collect(Collectors.toList());
```
Comment (ZekNikZ) [inline src/main/java/to/us/resume_builder/resume_components/category/BulletCategory.java]: Why not use `bullets.removeIf(b -> b.getID().equals(id))`?

File: src/main/java/to/us/resume_builder/resume_components/category/ExperienceCategory.java [hunk 2]
```diff
@@ -4,28 +4,75 @@
...
+     * @return a reference to the Experience if it is found, if not found return null.
+     */
+    Experience getExperienceByID(String id){
+        for (Experience e : experiences) {
+            if (e.getId().equals(id)) {
+                return e;
+            }
+        }
+        return null;
+    }
```
Comment (ZekNikZ) [inline src/main/java/to/us/resume_builder/resume_components/category/ExperienceCategory.java:40]: Another option to consider:

```java
return experiences.stream()
    .filter(c -> c.getId().equals(id))
    .findFirst()
    .orElse(null);
```

File: src/main/java/to/us/resume_builder/resume_components/category/ExperienceCategory.java [hunk 3]
```diff
@@ -4,28 +4,75 @@
...
+        return id;
     }
 
-    // TODO Implement
+    /**
+     * Removes the list item that matches the id.
+     * @param id
+     */
     public void removeExperience(String id){
+        experiences = experiences.stream().filter(e -> !e.getId().equals(id)).collect(Collectors.toList());
```
Comment (ZekNikZ) [inline src/main/java/to/us/resume_builder/resume_components/category/ExperienceCategory.java]: Why not use `experiences.removeIf(e -> e.getID().equals(id))`?

File: src/main/java/to/us/resume_builder/resume_components/category/HeaderCategory.java [hunk 4]
```diff
@@ -31,34 +31,66 @@ public HeaderCategory(String id) {
...
         this.email = email;
     }
 
-    public String getPhone_number() {
+    /**
+     * Returns the current String phone number
+     * @return the current String phone_number
+     */
+    public String getPhoneNumber() {
         return phone_number;
```
Comment (ZekNikZ) [inline src/main/java/to/us/resume_builder/resume_components/category/HeaderCategory.java]: Change the `phone_number` field to be camelCase (i.e. `phoneNumber`).

Comment (ZekNikZ) [reply src/main/java/to/us/resume_builder/resume_components/category/HeaderCategory.java]: (applies to whole class)

**Review by ZekNikZ — COMMENTED** [review]
Only a few efficiency and code style things. Otherwise looks fine.

File: src/main/java/to/us/resume_builder/resume_components/Experience.java [hunk 5]
```diff
@@ -44,61 +44,122 @@ public Experience(String id){
...
 
         // add new element to bullets
         bullets.add(new Bullet(id));
         return id;
     }
 
-    // TODO Implement
+    /**
+     * Removes the list item that matches the id
+     * @param id
```
Comment (ZekNikZ) [inline src/main/java/to/us/resume_builder/resume_components/Experience.java]: Illegal Javadoc syntax. Add parameter description.

File: src/main/java/to/us/resume_builder/resume_components/IBulletContainer.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * Create a new bullet for bullets list with a random generated id
+     * @return the id created for the new bullet
+     */
+    String addBullet();
+
+    /**
+     * Removes the list item that matches the id
+     * @param id
```
Comment (ZekNikZ) [inline src/main/java/to/us/resume_builder/resume_components/IBulletContainer.java]: Illegal Javadoc syntax. Add parameter description.

File: src/main/java/to/us/resume_builder/resume_components/ResumeComponent.java [hunk 7]
```diff
@@ -5,8 +5,16 @@
...
     protected final String id;
+
+    /**
+     * The visibility for the instance
+     */
     protected boolean visible = true;
 
+    /**
+     * sets the resume id for the instance
+     * @param id
```
Comment (ZekNikZ) [inline src/main/java/to/us/resume_builder/resume_components/ResumeComponent.java]: Illegal Javadoc syntax. Add parameter description.

File: src/main/java/to/us/resume_builder/resume_components/category/BulletCategory.java [hunk 8]
```diff
@@ -1,35 +1,83 @@
...
 
         // add new element to bullets
         bullets.add(new Bullet(id));
         return id;
     }
 
-    // TODO Implement
+    /**
+     * Removes the list item that matches the id
+     * @param id
```
Comment (ZekNikZ) [inline src/main/java/to/us/resume_builder/resume_components/category/BulletCategory.java]: Illegal Javadoc syntax. Add parameter description.

File: src/main/java/to/us/resume_builder/resume_components/category/Category.java [hunk 9]
```diff
@@ -16,27 +19,52 @@
...
+     * @param type the type that the category is
+     */
     protected Category(String id, CategoryType type){
         super(id);
         this.type = type;
     }
 
+    /**
+     * Returns the CategoryType type of the category.
+     * @return type
```
Comment (ZekNikZ) [inline src/main/java/to/us/resume_builder/resume_components/category/Category.java:37]: Illegal Javadoc syntax. Add return description.

File: src/main/java/to/us/resume_builder/resume_components/category/Category.java [hunk 10]
```diff
@@ -16,27 +19,52 @@
...
+     */
     public CategoryType getType() {
         return type;
     }
 
+    /**
+     * Returns the current String displayName.
+     * @return displayName
```
Comment (ZekNikZ) [inline src/main/java/to/us/resume_builder/resume_components/category/Category.java]: Illegal Javadoc syntax. Add return description.

File: src/main/java/to/us/resume_builder/resume_components/category/Category.java [hunk 11]
```diff
@@ -16,27 +19,52 @@
...
+     * Sets the String displayName.
+     * @param displayName the String to set to
+     */
     public void setDisplayName(String displayName) {
         this.displayName = displayName;
     }
 
+    /**
+     * Returns the current String name.
+     * @return name
```
Comment (ZekNikZ) [inline src/main/java/to/us/resume_builder/resume_components/category/Category.java]: Illegal Javadoc syntax. Add return description.

File: src/main/java/to/us/resume_builder/resume_components/category/ExperienceCategory.java [hunk 12]
```diff
@@ -4,28 +4,75 @@
...
+            String id = this.id + "." + rand.nextInt(1000);
+        } while (checkExperienceListID(id));
+        experiences.add(new Experience(id));
+        return id;
     }
 
-    // TODO Implement
+    /**
+     * Removes the list item that matches the id.
+     * @param id
```
Comment (ZekNikZ) [inline src/main/java/to/us/resume_builder/resume_components/category/ExperienceCategory.java]: Illegal Javadoc syntax. Add parameter description.

File: src/main/java/to/us/resume_builder/resume_components/category/HeaderCategory.java [hunk 13]
```diff
@@ -31,34 +31,66 @@ public HeaderCategory(String id) {
         super(id, CategoryType.HEADER);
     }
 
+    /**
+     * Returns the current String link
+     * @return link
```
Comment (ZekNikZ) [inline src/main/java/to/us/resume_builder/resume_components/category/HeaderCategory.java]: Illegal Javadoc syntax. Add return description.

File: src/main/java/to/us/resume_builder/resume_components/category/HeaderCategory.java [hunk 14]
```diff
@@ -31,34 +31,66 @@ public HeaderCategory(String id) {
...
+     */
     public String getLink() {
         return link;
     }
 
+    /**
+     * Sets the value of link
+     * @param link
```
Comment (ZekNikZ) [inline src/main/java/to/us/resume_builder/resume_components/category/HeaderCategory.java]: Illegal Javadoc syntax. Add parameter description.

File: src/main/java/to/us/resume_builder/resume_components/category/HeaderCategory.java [hunk 15]
```diff
@@ -31,34 +31,66 @@ public HeaderCategory(String id) {
...
+     */
     public void setLink(String link) {
         this.link = link;
     }
 
+    /**
+     * Returns the current String email
+     * @return email
```
Comment (ZekNikZ) [inline src/main/java/to/us/resume_builder/resume_components/category/HeaderCategory.java]: Illegal Javadoc syntax. Add return description.

File: src/main/java/to/us/resume_builder/resume_components/category/HeaderCategory.java [hunk 16]
```diff
@@ -31,34 +31,66 @@ public HeaderCategory(String id) {
...
+     */
     public String getEmail() {
         return email;
     }
 
+    /**
+     * Sets the value of email
+     * @param email
```
Comment (ZekNikZ) [inline src/main/java/to/us/resume_builder/resume_components/category/HeaderCategory.java]: Illegal Javadoc syntax. Add parameter description.

File: src/main/java/to/us/resume_builder/resume_components/category/HeaderCategory.java [hunk 17]
```diff
@@ -31,34 +31,66 @@ public HeaderCategory(String id) {
...
+     * Sets the String phone_number
+     * @param phone_number the new String to set to
+     */
+    public void setPhoneNumber(String phone_number) {
         this.phone_number = phone_number;
     }
 
+    /**
+     * Returns the current String address
+     * @return
```
Comment (ZekNikZ) [inline src/main/java/to/us/resume_builder/resume_components/category/HeaderCategory.java]: Illegal Javadoc syntax. Add return description.

File: src/main/java/to/us/resume_builder/resume_components/category/HeaderCategory.java [hunk 18]
```diff
@@ -31,34 +31,66 @@ public HeaderCategory(String id) {
...
+     */
     public String getAddress() {
         return address;
     }
 
+    /**
+     * Sets the String address
+     * @param address
```
Comment (ZekNikZ) [inline src/main/java/to/us/resume_builder/resume_components/category/HeaderCategory.java]: Illegal Javadoc syntax. Add parameter description.

File: src/main/java/to/us/resume_builder/resume_components/category/TextCategory.java [hunk 19]
```diff
@@ -2,16 +2,31 @@
...
+     * Creates an instance of TextCategory with id
+     * @param id the ID for this instance of Category
+     */
     public TextCategory(String id) {
         super(id, CategoryType.TEXT);
     }
 
+    /**
+     * Returns the current String text
+     * @return text
```
Comment (ZekNikZ) [inline src/main/java/to/us/resume_builder/resume_components/category/TextCategory.java]: Illegal Javadoc syntax. Add return description.

**Review by ZekNikZ — COMMENTED** [review]
Javadoc comments need updating

File: src/main/java/to/us/resume_builder/resume_components/category/ExperienceCategory.java (hunk 2 above)
Comment (AshleyLuEllen) [reply src/main/java/to/us/resume_builder/resume_components/category/ExperienceCategory.java:40]: Done

File: src/main/java/to/us/resume_builder/resume_components/category/BulletCategory.java (hunk 1 above)
Comment (AshleyLuEllen) [reply src/main/java/to/us/resume_builder/resume_components/category/BulletCategory.java]: Done

File: src/main/java/to/us/resume_builder/resume_components/category/ExperienceCategory.java (hunk 3 above)
Comment (AshleyLuEllen) [reply src/main/java/to/us/resume_builder/resume_components/category/ExperienceCategory.java]: Done

File: src/main/java/to/us/resume_builder/resume_components/category/HeaderCategory.java (hunk 4 above)
Comment (AshleyLuEllen) [reply src/main/java/to/us/resume_builder/resume_components/category/HeaderCategory.java]: Done

File: src/main/java/to/us/resume_builder/resume_components/Experience.java (hunk 5 above)
Comment (AshleyLuEllen) [reply src/main/java/to/us/resume_builder/resume_components/Experience.java]: Done

**Review by brooklynnStone1 — APPROVED** [review]

**Comment by AshleyLuEllen** [comment]
> Javadoc comments need updating

All have been done

**Review by ZekNikZ — APPROVED** [review]

**Review by nnazo — APPROVED** [review]

**Pull request merged**